File: backend/app/v1/schemas/job.py
"""
Pydantic schemas for Job-related data transfer.
"""

import uuid
from datetime import datetime

from app.v1.schemas.department import DepartmentRead
from app.v1.schemas.job_stage import JobStageConfigRead
from app.v1.schemas.skill import SkillRead
from pydantic import BaseModel, ConfigDict


class JobBase(BaseModel):
    """
    Base schema for Job data with shared attributes.
    """

    title: str
    vacancy: int | None = None
    department_id: uuid.UUID | None = None
    jd_text: str | None = None
    jd_json: dict | None = None
    is_active: bool = True
    passing_threshold: float = 65.0
    custom_extraction_fields: list[str] | None = None


class JobCreate(JobBase):
    """
    Schema for creating a new Job.
    """

    skill_ids: list[uuid.UUID] = []


class JobUpdate(BaseModel):
    """
    Schema for updating an existing Job.
    """

    title: str | None = None
    vacancy: int | None = None
    department_id: uuid.UUID | None = None
    jd_text: str | None = None
    jd_json: dict | None = None
    is_active: bool | None = None
    skill_ids: list[uuid.UUID] | None = None
    passing_threshold: float | None = None
    custom_extraction_fields: list[str] | None = None


class JobStatusUpdate(BaseModel):
    """
    Schema for updating only the job's active status.
    """

    is_active: bool


class JobVersionMinimal(BaseModel):
    """
    Minimal schema for a Job version, showing only version number and its unique ID.
    """

    version_num: int
    id: uuid.UUID
    model_config = ConfigDict(from_attributes=True)


class JobVersionRead(BaseModel):
    """
    Schema for reading full Job version snapshot data.
    """

    id: uuid.UUID
    job_id: uuid.UUID
    version_number: int
    title: str
    vacancy: int | None = None
    jd_text: str | None = None
    jd_json: dict | None = None
    custom_extraction_fields: list[str] | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class JobRead(JobBase):
    """
    Schema for reading Job data, including database-generated fields.
    """

    id: uuid.UUID
    version: int = 1
    total_versions: int = 0
    job_versions: list[JobVersionMinimal] = []
    created_by: uuid.UUID
    created_at: datetime
    department_name: str | None = None
    department: DepartmentRead | None = None
    skills: list[SkillRead] = []
    stages: list[JobStageConfigRead] = []
    decision_summary: dict | None = None
    automated_screening_summary: dict | None = None
    total_candidates: int | None = None
    current_session_candidates: int | None = None
    activity_sessions: list[JobActivitySession] | None = None

    model_config = ConfigDict(from_attributes=True)


class JobsListRead(BaseModel):
    """
    Schema for a paginated list of jobs.
    """

    data: list[JobRead]
    total: int
    global_decision_summary: dict | None = None
    global_screening_summary: dict | None = None

class JobActivitySession(BaseModel):
    """
    Schema for a single job activation session.
    Represents a period where the job was active.
    """

    session_id: int
    start_date: datetime
    end_date: datetime | None = None
    candidate_count: int
    is_current: bool

    model_config = ConfigDict(from_attributes=True)


class JobActivityHistoryResponse(BaseModel):
    """
    Response schema for job activity history.
    """

    job_id: uuid.UUID
    total_candidates: int
    sessions: list[JobActivitySession]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/job.ts
import type { JobStageConfig } from "@/types/stage";
import type { JobPositionRead } from "@/types/jobPosition";
import type { JobCreateFormValues, JobUpdateFormValues } from "@/schemas/job";
import type { JobPriorityRead } from "@/types/jobPriority";
import type { AssociateRead } from "@/types/associate";

/**
 * Represents a job posting in the hiring platform.
 * Contains job details, description, status information, and workflow metadata.
 */
export interface Job {
  /** Unique identifier for the job */
  id: string;
  /** Job title or position name */
  title: string;
  /** UUID of the department this job belongs to */
  department_id: string | null;
  /** Resolved department name (read-only, from relationship) */
  department_name: string | null;
  /** Full department object returned from the API */
  department?: { id: string; name: string; description: string | null } | null;
  /** Job description in plain text format */
  jd_text: string | null;
  /** Job description in structured JSON format */
  jd_json: Record<string, unknown> | null;
  /** Whether the job is currently accepting applications */
  is_active: boolean;
  /** Optional custom extraction fields used during resume parsing */
  custom_extraction_fields?: string[] | null;
  /** Threshold score (0-100) for considering a candidate as 'pass' */
  passing_threshold: number;
  question_bank_passing_threshold: number;
  /** Current active version number */
  version?: number;
  /** Total number of saved versions */
  total_versions?: number;
  /** Version history metadata */
  job_versions?: JobVersionMinimal[];
  /** Configured interview stages */
  stages?: JobStageConfig[];
  /** Backend-provided decision summary for the job */
  decision_summary?: JobDecisionSummary | null;
  /** Real-time automated screening summary from the backend */
  automated_screening_summary?: Record<string, any> | null;
  /** ID of the user who created the job posting */
  created_by: string;
  /** Timestamp when the job was created */
  created_at: string;
  /** Skills linked to the job */
  skills: { id: string; name: string; description?: string; default_weightage?: number }[];
  /** Associates linked to the job */
  associates?: AssociateRead[];
  job_skill_weightages?: Record<string, number> | null;

  total_candidates: number;
  current_session_candidates: number;
  activity_sessions?: JobActivitySession[] | null;
  vacancy: number | null;
  priority_id: string | null;
  associate_reminder_hours: number | null;
  priority_start_date: string | null;
  priority_end_date: string | null;
  priority?: JobPriorityRead | null;
  position_id: string;
  position?: JobPositionRead | null;
  processing_version?: number | null;
  message?: string | null;
  task_file_path?: string | null;
  task_skills?: string[] | null;
  send_ai_evaluation_to_associate: boolean;
}

export type JobCreate = JobCreateFormValues
export type JobUpdate = JobUpdateFormValues

export interface JobDecisionSummary {
  job_id: string;
  total_candidates: number;
  approved_count: number;
  reject_count: number;
  maybe_count: number;
  undecided_count: number;
}

export interface JobActivitySession {
  session_id: number;
  start_date: string;
  end_date: string | null;
  candidate_count: number;
  is_current: boolean;
  approved_count: number,
  rejected_count: number,
  pending_count: number
}

/**
 * Minimal job version metadata returned by the backend.
 */
export interface JobVersionMinimal {
  id: string;
  version_num: number;
}

export interface JobVersionDetail {
  id: string;
  job_id: string;
  version_number: number;
  title: string;
  jd_text: string | null;
  jd_json: Record<string, unknown> | null;
  custom_extraction_fields: string[] | null;
  created_at: string;
}

/**
 * Paginated response for job lists including global dashboard metrics.
 */
export interface JobsListResponse {
  data: Job[];
  total: number;
  global_decision_summary: Record<string, any> | null;
  global_screening_summary: Record<string, any> | null;
}


/**
 * Represents a job title.
 */
export interface JobTitle {
  id: string;
  title: string;
}

export interface JobTask {
  task_file_path: string | null;
  task_skills: string[] | null;
}

export interface DeleteJobTaskResponse {
  message: string;
}

export interface JobTitleVariant {
  job_id: string;
  position_id: string;
  position_name: string;
  is_active: boolean;
}

export interface JobTitleGroup {
  title: string;
  variants: JobTitleVariant[];
}

export interface JobTitlesGroupedListResponse {
  data: JobTitleGroup[];
}

/**
 * Candidate statistics for a specific job.
 */
export interface JobCandidatesStats {
  job_id: string;
  job_title: string;
  department?: string;
  candidate_count: number;
}

export interface JobPipelineStats {
  stage?: string;
  job_names?: string[];
  [jobTitle: string]: any;
}
